src/lib/certificate-manager.ts

```diff
@@ -0,0 +1,261 @@
+
+			// Create a GeneralSubtree containing the GeneralName
+			// According to the RFC: GeneralSubtree ::= SEQUENCE {
+			//   base GeneralName, minimum [0] BaseDistance DEFAULT 0, maximum [1] BaseDistance OPTIONAL }
+			return asn1.create(
+				asn1.Class.UNIVERSAL,
+				asn1.Type.SEQUENCE,
+				true,
+				[ generalName ]
+				// The minimum value is DEFAULT 0, so we can omit it for simplicity
+				// The maximum value is OPTIONAL and we don't need it
+			);
+		} );
+
+	// Build the NameConstraints structure
+	const nameConstraintsComponents: Array< forge.asn1.Asn1 > = [];
+
+	// Add permitted subtrees (tag 0)
+	nameConstraintsComponents.push(
+		asn1.create( asn1.Class.CONTEXT_SPECIFIC, 0, true, domainsToSequence( domains ) )
+	);
+
+	// Create the full NameConstraints structure
+	// NameConstraints ::= SEQUENCE { permittedSubtrees [0] GeneralSubtrees OPTIONAL,
+	//                                excludedSubtrees [1] GeneralSubtrees OPTIONAL }
+	const nameConstraintsValue = asn1.create(
+		asn1.Class.UNIVERSAL,
+		asn1.Type.SEQUENCE,
+		true,
+		nameConstraintsComponents
+	);
+
+	return {
+		id: '2.5.29.30', // nameConstraints OID
+		critical: true,
+		value: nameConstraintsValue,
+	};
+}
+
+// Certificate configuration
+const CA_NAME = 'WordPress Studio CA';
+const CA_CERT_VALIDITY_DAYS = 3650; // 10 years
+const SITE_CERT_VALIDITY_DAYS = 825; // a little over 2 years
+const CERT_DIRECTORY = getUserDataCertificatesPath();
+const CA_CERT_PATH = path.join( CERT_DIRECTORY, 'studio-ca.crt' );
+const CA_KEY_PATH = path.join( CERT_DIRECTORY, 'studio-ca.key' );
+
+// Make sure the certificates directory exists
+if ( ! fs.existsSync( path.join( CERT_DIRECTORY, 'domains' ) ) ) {
+	fs.mkdirSync( path.join( CERT_DIRECTORY, 'domains' ), { recursive: true } );
+}
+
+/**
+ * Generate a root CA certificate if it doesn't exist
+ */
+export async function ensureRootCA(): Promise< { cert: string; key: string } > {
+	// If the certificate already exists, no need to generate a new one
+	if ( fs.existsSync( CA_CERT_PATH ) && fs.existsSync( CA_KEY_PATH ) ) {
+		return {
+			cert: fs.readFileSync( CA_CERT_PATH, 'utf8' ),
+			key: fs.readFileSync( CA_KEY_PATH, 'utf8' ),
+		};
+	}
+
+	console.log( 'Generating new root CA certificate...' );
+
+	const keys = forge.pki.rsa.generateKeyPair( 2048 );
+	const cert = forge.pki.createCertificate();
+
+	cert.publicKey = keys.publicKey;
+	cert.serialNumber = crypto.randomBytes( 20 ).toString( 'hex' );
+	const now = new Date();
+	cert.validity.notBefore = now;
+	cert.validity.notAfter = new Date( now.getTime() );
+	cert.validity.notAfter.setDate( now.getDate() + CA_CERT_VALIDITY_DAYS );
+	const attrs = [
+		{ name: 'commonName', value: CA_NAME },
+		{ name: 'countryName', value: 'US' },
+		{ name: 'organizationName', value: 'WordPress Studio' },
+		{ name: 'organizationalUnitName', value: 'Development' },
+	];
+	cert.setSubject( attrs );
+	cert.setIssuer( attrs );
+
+	const extensions = [
```

Comment:
```suggestion
	cert.setExtensions( [
```
Total nit, but this'd be nice.